Invalidate existing sessions for a given User when User.password is changed

ewdurbin opened the pull request:
Closes #10849.

ewdurbin commented on warehouse/accounts/services.py:
```diff
@@ -80,14 +82,19 @@ def __init__(self, session, *, ratelimiters=None, remote_addr, metrics):
         )
         self.remote_addr = remote_addr
         self._metrics = metrics
+        self.cached_get_user = functools.lru_cache()(self._get_user)
```
Side effect of scrambling with figuring out why I couldn't get the latest `password_date` value... but interestingly this may resolve a memory leak? https://rednafi.github.io/reflections/dont-wrap-instance-methods-with-functoolslru_cache-decorator-in-python.html

di commented on warehouse/accounts/models.py:
```diff
@@ -108,6 +108,15 @@ class User(SitemapMixin, db.Model):
+    @password.setter
+    def password(self, value):
+        self._password = value
+        self.password_date = sql.func.now()
```
I think we can commit the transaction early to get the new date and thus retain the existing DB triggers. I think it's probably desirable for those to continue to remain in the DB so we don't easily forget to update `password_date`.

ewdurbin replied:
I can't figure out a way to get the resulting trigger value to be accessible inside the request. Calling `tm.commit()` mid request crashes the request at the end since txn complete... savepoint() doesn't work...

ewdurbin replied:
trying sqlalchemy expire call... might work!

ewjoachim replied:
... Ok, today is the day where my comments land 30s late.

ewdurbin replied:
yah, absolutely not ideal, but this is no longer necessary with https://github.com/pypa/warehouse/pull/10859/commits/8309b0057b0677000a577de4d81facd39a23653a.

for clarity, sql.func.now() gets translated to `now()` on insert, it's not calling the function.

ewjoachim replied:
(Yeah 👍  but to explicit my comment, if we had forced a commit at that point it would have been doing a SQL query in the setter)

ewdurbin replied:
ah, I wasn't proposing a commit in the setter, I was saying I had attempted all of these things in the view.

ewdurbin replied:
specifically, where refresh came to the rescue: https://github.com/pypa/warehouse/pull/10859/commits/8309b0057b0677000a577de4d81facd39a23653a#diff-55d42e8160b62924651b14b5c3649c6511c24a79876c8447cdad028da727d14fR364

di commented on warehouse/sessions.py:
```diff
@@ -145,6 +146,15 @@ def record_auth_timestamp(self):
+        self[self._password_timestamp_key] = timestamp
+        self.changed()
+
+    def password_outdated(self, current_password_timestamp):
+        stored_password_timestamp = self.get(self._password_timestamp_key, 0)
+
+        return current_password_timestamp != stored_password_timestamp
```
I think this would have the effect of logging out all currently active sessions because their sessions lack the `_password_timestamp` key, and thus this comparison will fail for them.

ewdurbin replied:
Great catch. I was initially going to implement the password date check based on a >=, then realized equality was a better check.

I'll update the default to None and add a conditional.

ewdurbin replied:
fixed in https://github.com/pypa/warehouse/pull/10859/commits/61ce40b0b30cadf1a907af1d552299ad474015fd

ewdurbin commented:
Sanity check done locally:

- Switched back to `main`
- Logged in with a brand new session
- Switched to this branch
- Stayed logged in!
- Opened new window and changed password
- New session that changed password stayed logged in, Old session terminated.

di commented on warehouse/sessions.py:
```diff
@@ -145,6 +146,20 @@ def record_auth_timestamp(self):
+        self.changed()
+
+    def password_outdated(self, current_password_timestamp):
+        stored_password_timestamp = self.get(self._password_timestamp_key, None)
```
```suggestion
        stored_password_timestamp = self.get(self._password_timestamp_key)
```
nit

ewdurbin replied:
but explicit 😭

di commented on warehouse/sessions.py:
```diff
@@ -145,6 +146,20 @@ def record_auth_timestamp(self):
+            # we cannot say for sure, let it live its life.
+            return False
+
+        return current_password_timestamp != stored_password_timestamp
```
I like that the `!=` comparison also covers our bases against time-travelers.

ewdurbin replied:
![](https://media.giphy.com/media/103gOGGaHcBq3C/giphy.gif)
